pylean/pylean/__init__.py

```diff
@@ -0,0 +1,37 @@
+import pexpect
+import json
+
+import os
+
+class LeanServer:
+    def __init__(self):
+        path_to_repl = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+
+        self.proc = pexpect.spawn(
+            "lake env lean --run REPL/Main.lean", cwd=path_to_repl, encoding="utf-8"
+        )
+
+    def run_code(self, code, env=None, verbose=False):
+        if env:
+            command = (
+                '{ "cmd" : "' + repr(code)[1:-1] + f'", "env" : {env}' + " }"
+            )  # [1:-1] removes single quotes
+        else:
+            command = (
+                '{ "cmd" : "' + repr(code)[1:-1] + '" }'
+            )  # [1:-1] removes single quotes
+
+        if verbose:
+            print(command)
+        self.proc.sendline(command)
+        self.proc.expect_exact(command + "\r\n")
+        self.proc.sendline()
+        self.proc.expect_exact("\r\n")
+        try:
+            index = self.proc.expect('env": \d+\}', timeout=20)
```

Comment:
Can you `expect` the whole line here and parse it as json, rather than parsing using regex?

Reply:
It can't because there are multiline outputs.

There seems to be no guarantee in the Lean code that `env` will appear at the last, so this is not robust.

Per my test of `repl` and [this comment in the code](https://github.com/leanprover-community/repl/blob/6d5b2c3dbfe7f8619d264bbc473ae13ac4aacf17/REPL/Main.lean#L317), `repl` is using `\r\n\r\n` as delimiters for requests/responses, just like in HTTP protocol, this is probably the only way to expect here, except to use a loop to expect more lines until the response becomes a valid JSON, which would wastes many JSON parse attempts, or we have to deal with these using fragile regexes.

My implementation inspired by this PR but expecting `\r\n\r\n` protocol is [here](https://github.com/utensil/lean4_jupyter/blob/main/lean4_jupyter/repl.py), notably, it turns off echo so there is less expecting.

For the stability of parsing `repl`'s output, I hope this `\r\n\r\n` protocol could be explicitly documented.